### PR TITLE
Initial implementation of cross-compilation to Knuth's MMIX

### PR DESCRIPTION
**Pull request by siraben**
###### Motivation for this change

Let's make Nix one of the first Linux package managers to have cross-compile support for [MMIX](https://en.wikipedia.org/wiki/MMIX)!

The following command cross-builds GNU Hello and runs it in Knuth's `mmixware` simulator for MMIX showing that indeed it works.
```sh
$ nix-shell -E 'with import ./. { }; runCommand "" { buildInputs = [ pkgsCross.mmix.hello mmixware ]; } ""' \                            
            --run 'mmix $(which hello)'
Hello, world!
```

Some variations I'm trying (failed ones are strikethroughed)
- [x] `mmix-unknown-mmixware` `libc = newlib`
- [x] ~`mmix-unknown-linux-uclib`~
- [x] ~`mmix-unknown-elf`~
- [x] ~`mmix-none-elf`~
- [ ] `mmix-elf`
- [x] `mmix`

Looks like we will only be able to use `newlib`, as it has `mmixware` [support](https://sourceware.org/git/?p=newlib-cygwin.git;a=tree;f=newlib/libc/sys/mmixware).

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [x] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/tools/mmixware/default.nix [hunk 1]
```diff
@@ -0,0 +1,42 @@
+{ stdenv, fetchurl, tetex }:
+
+stdenv.mkDerivation {
+  name = "mmixware";
```
Comment (SuperSandro2000) [inline pkgs/development/tools/mmixware/default.nix:4]: version number missing.

File: pkgs/development/tools/mmixware/default.nix [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+
+  src = fetchurl {
+    url = "https://www-cs-faculty.stanford.edu/~knuth/programs/mmix.tar.gz";
```
Comment (SuperSandro2000) [inline pkgs/development/tools/mmixware/default.nix:7]: Are there versioned downloads?

File: pkgs/development/tools/mmixware/default.nix [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+  unpackPhase = ''
+    mkdir src
+    unpackFile $src src
+  '';
+  srcRoot = "src";
+
+  hardeningDisable = [ "format" ];
+
+  patchPhase = ''
+    substituteInPlace Makefile --replace 'rm abstime.h' ""
```
Comment (jtojnar) [inline pkgs/development/tools/mmixware/default.nix:19]: Overriding `patchPhase` breaks `patches` field. Use `postPatch` instead.

File: pkgs/development/tools/mmixware/default.nix [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+  '';
+
+  nativeBuildInputs = [ tetex ];
+  enableParallelBuilding = true;
+
+  makeFlags = [ "all" "doc" ];
+
+  installPhase = ''
+    mkdir -p $out/share/doc
```
Comment (jtojnar) [inline pkgs/development/tools/mmixware/default.nix:28]: When overriding `installPhase`, you should also provide `runHook preInstall` and `runHook postInstall`.

File: pkgs/development/tools/mmixware/default.nix (hunk 1 above)
Comment (Ericson2314) [reply pkgs/development/tools/mmixware/default.nix:4]:
```suggestion
  pname = "mmixware";
```

File: lib/systems/examples.nix [hunk 5]
```diff
@@ -109,6 +109,10 @@ rec {
     platform = platforms.riscv-multiplatform "32";
   };
 
+  mmix = {
+    config = "mmix";
```
Comment (Ericson2314) [inline lib/systems/examples.nix:113]: I think I would rather that this be
```suggestion
    config = "mmix-none";
```
which is a bit more of a normal form?

Comment (siraben) [reply lib/systems/examples.nix:113]: I think I tried this and GCC couldn't build. I'll try and see.

Comment (Ericson2314) [reply lib/systems/examples.nix:113]: Actually, How does the GNU hello example work? Is this using glibc? In that case we should call it:
```suggestion
    config = "mmix-unknown-linux-gnu";
```

Comment (Ericson2314) [reply lib/systems/examples.nix:113]: Oh I see https://github.com/bminor/newlib/blob/master/newlib/libc/sys/mmixware/write.c ; so the emulator has some very basic syscalls. The syscall interface is what the OS triple comes from (it's the interface not implementation that counts).

I'll keep on researching.

Comment (Ericson2314) [reply lib/systems/examples.nix:113]: OK `mmix-knuth-mmixware` is a thing that goes up, where `mmixware` refers to that syscall interface. Can you try `mmix-unknown-mmixware`? (The vendor field is pretty stupid so I rather stick with `unknown` if we can.)

Comment (Ericson2314) [reply lib/systems/examples.nix:113]:
```suggestion
    config = "mmix-unknown-mmixware";
```

Comment (siraben) [reply lib/systems/examples.nix:113]: Ok, trying.

Comment (siraben) [reply lib/systems/examples.nix:113]: error: Target specification with 3 components is ambiguous
(use '--show-trace' to show detailed location information)

Comment (siraben) [reply lib/systems/examples.nix:113]: I see, I need to adjust `parse.nix` as needed.

Comment (Ericson2314) [reply lib/systems/examples.nix:113]: And note the special case should be `mmixware` not `mmix`.

Comment (siraben) [reply lib/systems/examples.nix:113]: Done.

**Comment by SuperSandro2000** [comment]
Result of `nixpkgs-review pr 102766` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>mmixware</li>
  </ul>
</details>

File: lib/systems/parse.nix [hunk 6]
```diff
@@ -393,6 +395,8 @@ rec {
         then { cpu = elemAt l 0; vendor = elemAt l 1; kernel = "wasi";                       }
       else if (elemAt l 2 == "redox")
         then { cpu = elemAt l 0; vendor = elemAt l 1; kernel = "redox";                      }
+      else if (elemAt l 2 == "mmixware")
+        then { cpu = elemAt l 0; vendor = elemAt l 1; kernel = "none";                       }
```
Comment (Ericson2314) [inline lib/systems/parse.nix:399]: mmixware
```suggestion
        then { cpu = elemAt l 0; vendor = elemAt l 1; kernel = "mmixware";                   }
```

Comment (siraben) [reply lib/systems/parse.nix:399]:
```
error: Unknown kernel: mmixware
```

**Comment by Ericson2314** [comment]
There are these two `none`s left over. Note you will also need to make `mmixware` a "kernel". (we should really rename "kernel" to something better!!)

**Comment by siraben** [comment]
Where's the two `none`s? I found one in `doubles.nix`

**Comment by kisik21** [comment]
Please clarify: can this architecture run the Linux kernel or is it just bare-metal for now?

**Comment by siraben** [comment]
Currently bare metal. I'll adjust the top comment with variations tried.

**Comment by siraben** [comment]
It appears possible to boot the Linux kernel somehow, see http://mmix.cs.hm.edu/examples/Linux/index.html

File: pkgs/development/tools/mmixware/default.nix [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+
+  postPatch = lib.concatStringsSep "\n" [
+    "substituteInPlace Makefile --replace 'rm abstime.h' ''"
+  (lib.optionalString stdenv.targetPlatform.isMmix ''
+    mmotype abstime
+    substituteInPlace Makefile --replace './abstime' 'mmix abstime'
+  '')]
+  ;
+
+  nativeBuildInputs = [ tetex ] ++ lib.optional stdenv.targetPlatform.isMmix mmixware;
```
Comment (kisik21) [inline pkgs/development/tools/mmixware/default.nix:25]: `s/lib/stdenv.lib/g`

Comment (kisik21) [inline pkgs/development/tools/mmixware/default.nix:25]: Also, is this derivation... self-referencing? in L25 there's an undefined variable `mmixware`...

**Review by vikanezrimaya — CHANGES_REQUESTED** [review]
I am unable to reproduce the `nix-shell` example above. Maybe I'm doing something wrong?

File: pkgs/development/tools/mmixware/default.nix (hunk 7 above)
Comment (siraben) [reply pkgs/development/tools/mmixware/default.nix:25]: Oops I accidentally committed this.

**Comment by siraben** [comment]
@kisik21 Please try again.

**Comment by kisik21** [comment]
@siraben yay, it's compiling!

**Comment by siraben** [comment]
@Ericson2314 I added this target locally, and it works.
```nix
  mmix-embedded = {
    config = "mmix";
    libc = "newlib";
  };
```

However GNU hello's diff seems to be very small. Are they really different targets?
```diff
diff --git a/hello_mmixware.txt b/hello_mmixware-embedded.txt
index 99e4e08..0ca8e94 100644
--- a/hello_mmixware.txt
+++ b/hello_mmixware-embedded.txt
@@ -1,4 +1,4 @@
-00000000: 9809 0101 5fa2 a96c 9801 0002 0000 0000  ...._..l........
+00000000: 9809 0101 5fa3 ea56 9801 0002 0000 0000  ...._..V........
 00000010: 0000 0100 6000 0000 0000 0000 0000 0000  ....`...........
 00000020: 0002 9758 0000 0000 0002 983c 2000 0000  ...X.......< ...
 00000030: 0000 0000 2000 0000 0000 1480 e3ff 0020  .... .......... 
```

**Review by vikanezrimaya — APPROVED** [review]
Built and run the provided example successfully!

**Comment by kisik21** [comment]
It looks like ./lib/tests/release.nix needs updating to not see the new architecture as an error.

**Comment by siraben** [comment]
Looks like 

> It looks like ./lib/tests/release.nix needs updating to not see the new architecture as an error.

What do you mean by this?

**Comment by kisik21** [comment]
```
$ nix-build --arg pkgs 'import ./. {}' ./lib/tests/release.nix
warning: unknown setting 'extra-sandbox-paths'
error: --- ThrownError ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ nix-build
[{"expected":["aarch64-darwin","aarch64-genode","aarch64-linux","aarch64-none","arm-none","armv5tel-linux","armv6l-linux","armv6l-none","armv7a-darwin","armv7a-linux","armv7l-linux","avr-none","i686-cygwin","i686-darwin","i686-freebsd","i686-genode","i686-linux","i686-netbsd","i686-none","i686-openbsd","i686-windows","js-ghcjs","mipsel-linux","msp430-none","powerpc-none","powerpc64le-linux","riscv32-linux","riscv32-none","riscv64-linux","riscv64-none","vc4-none","wasm32-wasi","wasm64-wasi","x86_64-cygwin","x86_64-darwin","x86_64-freebsd","x86_64-genode","x86_64-linux","x86_64-netbsd","x86_64-none","x86_64-openbsd","x86_64-redox","x86_64-solaris","x86_64-windows"],"name":"testall","result":["aarch64-darwin","aarch64-genode","aarch64-linux","aarch64-none","arm-none","armv5tel-linux","armv6l-linux","armv6l-none","armv7a-darwin","armv7a-linux","armv7l-linux","avr-none","i686-cygwin","i686-darwin","i686-freebsd","i686-genode","i686-linux","i686-netbsd","i686-none","i686-openbsd","i686-windows","js-ghcjs","mipsel-linux","mmix-mmixware","msp430-none","powerpc-none","powerpc64le-linux","riscv32-linux","riscv32-none","riscv64-linux","riscv64-none","vc4-none","wasm32-wasi","wasm64-wasi","x86_64-cygwin","x86_64-darwin","x86_64-freebsd","x86_64-genode","x86_64-linux","x86_64-netbsd","x86_64-none","x86_64-openbsd","x86_64-redox","x86_64-solaris","x86_64-windows"]}]
```

Passing that through `jq '.[0].result - .[0].expected'` gets me `["mmix-mmixware"]` - it means that the lib test suite doesn't expect to see the new double in the list of available doubles.

**Comment by SuperSandro2000** [comment]
Result of `nixpkgs-review pr 102766` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>mmixware</li>
  </ul>
</details>

**Comment by siraben** [comment]
@kisik21 I'm not sure what files I have to change to resolve that error.

**Comment by FRidh** [comment]
You need to add the new arch to `lib/tests/systems.nix`. I suggest having a look at the file's log, because this issue happened more often when new architectures were added.

**Comment by Ericson2314** [comment]
@kisik21 

> Please clarify: can this architecture run the Linux kernel or is it just bare-metal for now?

The simulator supports some basic system calls. So it's like having a basic OS that just read writes out to serial ports or something.

**Review by Ericson2314 — APPROVED** [review]
Looks good!

**Pull request merged**